File: apps/analytics-engine/app/infrastructure/notification/discord_notifier.py
"""Discord notifier: sends notifications via Discord webhook URLs."""
from __future__ import annotations

from typing import Any

import structlog

from ...domain.value_objects.notification import (
    NotificationEvent,
    NotificationSeverity,
)
from ...application.ports.notifier import Notifier

log = structlog.get_logger()

_COLORS = {
    NotificationSeverity.INFO: 0x00FF00,
    NotificationSeverity.WARN: 0xFFA500,
    NotificationSeverity.CRITICAL: 0xFF0000,
}


class DiscordNotifier(Notifier):
    def __init__(self, webhook_url: str, http_client: Any | None = None) -> None:
        self._url = webhook_url
        self._http = http_client

    async def send(self, event: NotificationEvent) -> None:
        if self._http is None:
            try:
                import httpx
                self._http = httpx.AsyncClient()
            except ImportError:
                log.warning("httpx_not_available", discord_skipped=True)
                return

        embed = {
            "title": event.title,
            "description": event.message,
            "color": _COLORS.get(event.severity, 0x808080),
            "timestamp": event.timestamp.isoformat(),
        }
        if event.symbol:
            embed["fields"] = [{"name": "Symbol", "value": event.symbol, "inline": True}]

        payload = {"embeds": [embed]}

        try:
            resp = await self._http.post(
                self._url, json=payload, timeout=10
            )
            if resp.status_code not in (200, 204):
                log.warning(
                    "discord_webhook_error",
                    status=resp.status_code,
                    body=resp.text[:200],
                )
        except Exception as e:
            log.warning("discord_send_failed", error=str(e))

File: apps/data-engine/src/infrastructure/notification/discord-notifier.ts
const COLORS: Record<string, number> = {
  INFO: 0x00ff00,
  WARN: 0xffa500,
  CRITICAL: 0xff0000,
}

const log = (m: string): void => {
  // eslint-disable-next-line no-console
  console.log(m)
}

export interface DiscordPayload {
  event_type: string
  severity: string
  title: string
  message: string
  symbol?: string
  timestamp: string
}

/**
 * Sends a notification to a Discord channel via webhook URL.
 * Fire-and-forget: failures are logged but not thrown.
 */
export async function sendDiscord(payload: DiscordPayload): Promise<void> {
  const url = process.env.DISCORD_WEBHOOK_URL
  if (!url) return

  const embed: Record<string, unknown> = {
    title: payload.title,
    description: payload.message,
    color: COLORS[payload.severity] ?? 0x808080,
    timestamp: payload.timestamp,
  }
  if (payload.symbol) {
    embed.fields = [{ name: "Symbol", value: payload.symbol, inline: true }]
  }

  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ embeds: [embed] }),
    })
    if (!res.ok && res.status !== 204) {
      const body = await res.text()
      log(`[discord] webhook error ${res.status}: ${body.slice(0, 200)}`)
    }
  } catch (err) {
    log(`[discord] send failed: ${String(err)}`)
  }
}
